InformationScripting/src/queries/QueryRegistry.cpp

```diff
@@ -0,0 +1,112 @@
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "QueryRegistry.h"
+
+#include "../queries/AstQuery.h"
+#include "../queries/AstNameFilter.h"
+#include "../queries/NodePropertyAdder.h"
+#include "../queries/ScriptQuery.h"
+
+namespace InformationScripting {
+
+QueryRegistry& QueryRegistry::instance()
+{
+	static QueryRegistry instance;
+	return instance;
+}
+
+Query* QueryRegistry::buildQuery(const QString& command, Model::Node* target, QStringList args)
+{
+	if (auto constructor = constructors_[command])
+		return constructor(target, args);
+	if (args.size() > 1 && args[0] == "=")
+	{
+		// TODO we need some way to specify a condition on the node.
+		// Or eventually decide that we don't allow condition in the property adder
+		return new NodePropertyAdder(command, args[1]);
+	}
+	if (auto script = tryBuildQueryFromScript(command, args))
+		return script;
+	return nullptr;
+}
+
+QueryRegistry::QueryRegistry()
+{
+	registerDefaultQueries();
+	registerDefaultScriptLocations();
+}
+
+void QueryRegistry::registerDefaultQueries()
+{
+		registerQueryConstructor("classes", [](Model::Node* target, QStringList args) {
+			return new AstQuery(AstQuery::QueryType::Classes, target, args);
+		});
+		registerQueryConstructor("methods", [](Model::Node* target, QStringList args) {
+			return new AstQuery(AstQuery::QueryType::Methods, target, args);
+		});
+		registerQueryConstructor("bases", [](Model::Node* target, QStringList args) {
+			return new AstQuery(AstQuery::QueryType::BaseClasses, target, args);
+		});
+		registerQueryConstructor("toClass", [](Model::Node* target, QStringList args) {
+			return new AstQuery(AstQuery::QueryType::ToClass, target, args);
+		});
+		registerQueryConstructor("callgraph", [](Model::Node* target, QStringList args) {
+			return new AstQuery(AstQuery::QueryType::CallGraph, target, args);
+		});
+		registerQueryConstructor("filter", [](Model::Node*, QStringList args) {
+			Q_ASSERT(args.size());
+			if (args[0].contains("*"))
+			{
+				QString regexString = args[0];
+				regexString.replace("*", "\\w*");
+				return new AstNameFilter(Model::SymbolMatcher(new QRegExp(regexString)));
+			}
+			return new AstNameFilter(Model::SymbolMatcher(args[0]));
+		});
+		registerQueryConstructor("script", [](Model::Node*, QStringList args) {
+			QString scriptName = args.takeFirst();
+			return new ScriptQuery(QString("../InformationScripting/test/scripts/%1.py").arg(scriptName), args);
+		});
+}
+
+void QueryRegistry::registerDefaultScriptLocations()
+{
+	scriptLocations_ << "../InformationScripting/test/scripts/";
```

Comment:
I agree that things might require an extra step, but it's the correct thing to do. Imagine a release, binary-only distribution of Envision. Then there should be a directory outside of the source code that has the scripts.

To ease debugging, you might on occasion directly modify the script in DebugBuild and once you're happy with it, just copy it back to the source directory.